vllm/model_executor/models/phi_1_5.py

```diff
@@ -0,0 +1,274 @@
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        cache_events: Optional[List[torch.cuda.Event]],
+    ) -> torch.Tensor:
+        hidden_states = self.wte(input_ids)
+        for i in range(len(self.layers)):
+            if cache_events is None:
+                cache_event = None
+            else:
+                cache_event = cache_events[i]
+            layer = self.layers[i]
+            hidden_states = layer(
+                position_ids,
+                hidden_states,
+                kv_caches[i],
+                input_metadata,
+                cache_event,
+            )
+        return hidden_states
+
+
+class PhiForCausalLM(nn.Module):
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.phi = PhiModel(config)
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.linear = ColumnParallelLinear(
+            config.n_embd,
+            config.vocab_size,
+            gather_output=False,
+        )
+        self.sampler = Sampler(config.vocab_size)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        cache_events: Optional[List[torch.cuda.Event]],
+    ) -> SamplerOutput:
+        hidden_states = self.phi(input_ids, positions, kv_caches,
+                                 input_metadata, cache_events)
+        hidden_states = self.ln(hidden_states)
+        next_tokens = self.sampler(self.linear.weight, hidden_states,
+                                   input_metadata, self.linear.bias)
+        return next_tokens
+
+    _column_parallel_weights = [
+        "embed_in.weight", "embed_out.weight", "embed_out.bias", "fc1.weight",
+        "fc1.bias"
+    ]
+    _row_parallel_weights = ["out_proj.weight", "fc2.weight"]
+
+    def load_weights(self,
+                     model_name_or_path: str,
+                     cache_dir: Optional[str] = None,
+                     load_format: str = "auto",
+                     revision: Optional[str] = None):
+        tensor_model_parallel_rank = get_tensor_model_parallel_rank()
+        state_dict = self.state_dict()
+        for name, loaded_weight in hf_model_weights_iterator(
+                model_name_or_path, cache_dir, load_format, revision):
+            if "rotary_emb.inv_freq" in name:
+                t = torch.arange(self.config.n_positions, dtype=torch.float32)
+
+                freqs = torch.einsum("i,j -> ij", t, loaded_weight)
+                cos = freqs.cos()
+                sin = freqs.sin()
+                cache = torch.cat((cos, sin), dim=-1)
+
+                for i in range(len(self.phi.layers)):
+                    self.phi.layers[
+                        i].mixer.attn.rotary_emb.cos_sin_cache.copy_(cache)
```

Comment:
Because the checkpoint has `inv_freq` for each layer, the loop copies the cache `num_layers * num_layers` times in total.

```suggestion
                layer_idx = int(name.split(".")[1]) - 1
                self.phi.layers[layer_idx].mixer.attn.rotary_emb.cos_sin_cache.copy_(cache)
```